Add builtin functions Core._import, Core._using; implement import/using logic in Julia

xal-0 opened the pull request:
# Overview

Currently, `import` and `using` statements are compiled into calls to
`jl_toplevel_eval` with the Expr as an argument.  It can be useful in an
interactive environment to do import/using programmatically, for which
`eval(Expr(:import, ...))` or a macro is the only option at the moment (see for
example `InteractiveUtils.@activate`).

This PR adds two builtins, `_module_import ` and `_module_using`, with these
signatures, and removes `Expr(:import/:using, ...)` from the lowered IR:

```
_module_import(explicit::Bool, to::Module, ctx::Union{Expr, Nothing}, names::Expr...)
_module_using(to::Module, from::Expr{Symbol})
```

## Lowering example

`import` statements and `using X:` statements are lowered to `_module_import`
like so:

```
import A             => _module_import(true,  Main, nothing,          Expr(:., :A))
import A.b           => _module_import(true,  Main, nothing,          Expr(:., :A, :b))
import A.b as c      => _module_import(true,  Main, nothing,          Expr(:as, Expr(:., :A, :b), :c))
import A.B: C.d, e   => _module_import(true,  Main, Expr(:., :A, :B), Expr(:., :C, :d), Expr(:., :e))
import A.B: C.d as e => _module_import(true,  Main, Expr(:., :A, :B), Expr(:as, Expr(:., :C, :d), :e))
using  A.B: C.d, e   => _module_import(false, Main, Expr(:., :A, :B), Expr(:., :C, :d), Expr(:., :e))
```

Plain `using` statements become `_module_using`:

```
using A.B            => _module_using(Main, Expr(:., :A, :B))
```

Multiple comma-separated `using` or `import` paths are lowered to multiple calls to the appropriate builtin:
```
julia> Meta.@lower using A.B, C
:($(Expr(:thunk, CodeInfo(
1 ─       builtin Core._module_using(Main, $(QuoteNode(:($(Expr(:., :A, :B))))))
│         builtin Core._module_using(Main, $(QuoteNode(:($(Expr(:., :C))))))
│       $(Expr(:latestworld))
└──     return nothing
))))
```

# Alternative designs

Some alternatives to these builtins are worth considering.

The lowering pass could do even less work by inserting a call to
`_module_import` or `_module_using` that takes a single Expr argument for the
entire statement.

Another option is to do even more in lowering, introducing very low-level
builtins that wrap `jl_module_import` and `jl_module_using` directly.  The
lowering pass would expand the import path into direct calls to `Base.require`,
`Base.getproperty`, etc.  For example, `import A.B: C.d, e` could become
something like:
  
```
(block
  (= (ssavalue 1)
     (call (top getproperty)
           (call (top require) (thismodule) 'A)
           'B))
  (call (core module_import)
        (thismodule)
        (call (top getproperty) (ssavalue 1) 'C)
        'd)
  (call (core module_import) (thismodule) (ssavalue 1) 'd)
  (latestworld))
```

I like the idea of supporting calls to `_module_import` with a module for
`ctx` instead of an Expr, but this could be added to the proposed design too.

xal-0 committed:
Add builtin functions _module_import, _module_using

Removes the lowered IR forms Expr(:using, ...) and Expr(:import, ...).  `import`
and `using` expressions are now lowered to calls to `_module_import` like so:

  import A, B.c

    (block
      (call (core _module_import) (true) (thismodule) (null) (inert (|.| A)))
      (call (core _module_import) (true) (thismodule) (null) (inert (|.| B c)))
      (latestworld))

  import A.B: C.d, e as f

    (block
      (call (core _module_import) (true) (thismodule)
            (inert (|.| A B))
            (inert (|.| C d))
            (inert (as (|.| e) f))))

  using A.B: C.d

    (block
      (call (core _module_import) (false) (thismodule)
            (inert (|.| A B))
            (inert (|.| C d)))
      (latestworld))

Each comma-separated import is lowered to a call to the new _module_import
builtin, which takes an "explicit" flag (`using X:` is lowered to _module_import
with explicit=false), a "context" path (set to `nothing` for `import` without
`:`), and any number of Exprs of the form `(|.| ...)` or `(as (|.| ...) x)`
representing import paths under the context path.

Without `:`, using is lowered to the other new builtin:

    using A.B

    (block
      (call (core _module_using) (thismodule) (inert (|.| A B)))
      (latestworld))

Keno commented:
Generally positive on the direction.

For this particular implementation, I'll need to think about it in a bit more depth, but my initial impression is that this feels too "syntax oriented" for a builtin. I mean this in two senses. The first is that these builtins are taking `Expr`, which is a syntax data structure and not really touched anywhere else by the runtime system. The second is that it conflates the resolution of the name to be imported with the establishment of the link between them. I'd have expected something closer to:
```
_module_import(::Binding, ::Binding, imported::Bool)
_module_using(::Module, ::Module)
```
From this, there could then be higher level functions or macros that implement the syntax parts. The primary downside I see is that this approach would prevent batching of the world age update. However, I think that is possibly better handled with a generic world agree compression optimization in the runtime rather than trying to design the syntax around it.

vtjnash commented:
Agreed as to our eventual plan here. I think there are few phases, one (here) is keeping most of the existing code structure, but introducing a function form for it. Secondly we can move a lot of this C code into julia using some simpler builtins for the primitive bits. I think the tricky part of deciding on that form is dealing with the fact this is very much just interpreting the syntax, so it is hard to see what that julia function's arguments should be other than exactly the AST.

xal-0 committed:
Fix creating const bindings for `using A` in Main

xal-0 committed:
Fix REPL modules_to_be_loaded to detect import/using builtins

xal-0 committed:
Merge remote-tracking branch 'upstream/master' into special-form-builtins

xal-0 committed:
Move import/using logic to Julia; add low-level Core._import, _using

Renames _module_import -> _eval_import and _module_using -> _eval_using (we have
jl_module_using and jl_module_using in C and it's confusing enough already), and
moves the logic into Julia.

Since we need _eval_import very early in bootstrapping, boot.jl provides a
minimal version that understands only `import M: a, b, ...`.  Some imports in
code included by Base_compiler.jl has been updated to match this.

The Julia _eval_* functions call two new low-level builtins:
- Core._import calls jl_module_import with the normal 5 argument form, or
  jl_import_module for the special case of importing a package to get a
  PARTITION_KIND_CONST_IMPORT binding (3 argument form).
- Core._using is a direct wrapper around jl_module_using.

xal-0 committed:
Merge remote-tracking branch 'upstream/master' into special-form-builtins

xal-0 commented:
c638398 is a more-or-less direct translation of `_module_import` and `_module_using` into Julia, renamed to `_eval_import` and `_eval_using` because we already have `jl_module_import` and `jl_import_module` and it was a little confusing.  The low-level builtins are now:

>    `_import(to::Module, from::Module, asname::Symbol, [sym::Symbol, imported::Bool])`
> 
> With all five arguments, imports `sym` from module `from` into `to` with name
> `asname`.  `imported` is true for bindings created with `import` (set it to
> false for `using A: ...`).
> 
> With only the first three arguments, creates a binding for the module `from`
> with name `asname` in `to`.

>    `_using(to::Module, from::Module)`
> 
> Add `from` to the usings list of `to`.


`Core._using` is a direct wrapper around `jl_module_using`. `Core._import` either calls `jl_module_import` (first form),or `jl_import_module` (second form, installs a `PARTITION_KIND_CONST_IMPORT`). If we could remove `PARTITION_KIND_CONST_IMPORT` it could be deleted.

xal-0 committed:
Merge branch 'master' into special-form-builtins

vtjnash commented on stdlib/REPL/src/REPL.jl:
```diff
@@ -358,26 +358,31 @@ function check_for_missing_packages_and_run_hooks(ast)
-                    push!(mods, sym)
-                end
+    if ast.head == :call
+        if ast.args[1] == GlobalRef(Core, :_eval_import)
```
```suggestion
        if ast.args[1] === GlobalRef(Core, :_eval_import)
```

vtjnash commented on base/module.jl:
```diff
@@ -0,0 +1,156 @@
+function eval_import_path(at::Module, from::Union{Module, Nothing}, path::Expr, keyword::String)
+    isempty(path.args) && error("malformed import statement")
+
+    i = 1
```
```suggestion
    i::Int = 1
```
to prevent invalidations

vtjnash commented on base/module.jl:
```diff
@@ -0,0 +1,156 @@
+    delete_method(m)
+end
+
+function call_require(into::Module, name::Symbol)
+    # TODO: JL_TIMING(LOAD_IMAGE, LOAD_Require)
+    build_mode = Bool(Base.JLOptions().incremental) && Bool(ccall(:jl_generating_output, Cint, ()))
+    m = nothing
+    req_w = unsafe_load(cglobal(:jl_require_world, UInt))
+    cur_w, tls_w = get_world_counter(), tls_world_age()
+    if isdefined(Base, :require)
+        w = build_mode && req_w < tls_w ? req_w : cur_w
+        m = invoke_in_world(w, getglobal(Base, :require), into, name)
+    end
+    m isa Module || error("failed to load module $v")
+    m
+end
```
IIUC, this can use `Base._require_world_age[]` now (see loading.jl) to simplify this code

vtjnash commented on src/julia-syntax.scm:
```diff
@@ -2939,6 +2953,38 @@
+      `(block
+        (toplevel-only import)
+        ,.(if (eq? (caadr e) ':)
+              `((call (core _eval_import) (true) (thismodule)
```
nit: `(core)` is usually for builtin-provided (not changeable) things, while `(top)` is usually for our user-definable things (from Base). The added advantage with `(top)` is that different modules could then chose to define a different "top" module (for Core & its submodules it is specified to be Core, while for Base and other modules it defaults to Base)
```suggestion
              `((call (top _eval_import) (true) (thismodule)
```

xal-0 replied:
I was wondering what the better option was here. Ultimately I went with core only because we need `_eval_import` inside of `boot.jl` and Base is not defined until later. I'll figure out how to work around that

vtjnash commented on src/module.c:
```diff
@@ -1269,24 +1269,26 @@ static void module_import_(jl_task_t *ct, jl_module_t *to, jl_module_t *from, jl
+    jl_sym_t *name = asname ? asname : import->name;
+    // TODO: this is a bit race-y with what error message we might print
+    jl_binding_t *b = jl_get_module_binding(m, name, 1);
+    jl_binding_partition_t *bpart = jl_get_binding_partition(b, ct->world_age);
```
This appears to be an existing mistake in `module_import_`, but `ct->world_age` is not correct to use in this function (these functions apply to `jl_world_counter`, so they will operate incorrectly if they instead reference world_age), so we should make a separate issue to fix this before the v1.12 release

xal-0 replied:
I think we avoided this bug on master by forcing imports to go through `jl_toplevel_eval_in`, so we'd always be in the latest world:
https://github.com/JuliaLang/julia/blob/6d78a4ae91078cea43470f8cd6d895e0c9b3d5e7/src/toplevel.c#L1094-L1095

Also discovered I broke `import Package as M` while looking in to this, woops. At least we'll have a test for it now.

vtjnash commented on stdlib/REPL/src/REPL.jl:
```diff
@@ -358,26 +358,31 @@ function check_for_missing_packages_and_run_hooks(ast)
-                end
+    if ast.head == :call
+        if ast.args[1] == GlobalRef(Core, :_eval_import)
+            ctx = ast.args[4]
```
Need to check array length before accessing this (or use `isexpr(:call, 5)`)

xal-0 committed:
Apply suggestions from code review

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

xal-0 committed:
Remove call_require now that Base.require uses invoke_in_world

Co-authored-by: Jameson Nash <jameson@juliacomputing.com>

xal-0 committed:
Check Expr.args length in REPL import hook

Co-authored-by: Jameson Nash <jameson@juliacomputing.com>

xal-0 committed:
Fix `import Package as M`

xal-0 committed:
Fix jl_import_module world age

xal-0 committed:
Refer to _eval_using and _eval_import through top module

KristofferC commented on base/module.jl:
```diff
@@ -0,0 +1,144 @@
+        elseif v === :Base
+            m = Base
+        else
+            # TODO: JL_TIMING(LOAD_IMAGE, LOAD_Require)
```
Can use `Core.Compiler.@zone` now perhaps.

xal-0 committed:
Re-add timing zone for LOAD_Require

xal-0 committed:
Merge remote-tracking branch 'upstream/master' into special-form-builtins

IanButterworth commented on test/worlds.jl:
```diff
@@ -558,3 +558,12 @@ module C57316; import ..X57316.Y57316 as Z, .Z.Y57316 as W; end
+    Core._eval_import(true, @__MODULE__, nothing, Expr(:., :Random))
+end
+end
+@test_throws ErrorException("importing Random into M57965 conflicts with an existing global") M57965.f()s
```
I just noticed `M57965.f()s`. Note the `s` at the end.. I guess it's not a problem because the `f()` errors before hitting that..

oscardssmith commented:
is `Core._using` necessary to have as a builtin? Couldn't it have just been a wrapper over `Core._import`?

c42f commented:
Drive by comment: this is great, it's exactly what I needed to fix some janky use of `eval()` in JuliaLowering.jl, thank you!

(It also breaks JuliaLowering, but that's a price I'm very much willing to pay :laughing:)
